internal/app/referenceclient/tracer.go

```diff
@@ -104,7 +94,7 @@ func (t *WireTracer) Complete(trace tracer.Trace) {
 				if statusCode != 200 {
 					// If this is a unary request, then use the entire response body
 					// as the wire error details.
-					body, err := io.ReadAll(trace.Response.Body)
+					body, err := io.ReadAll(wrapper.buf)
```

Comment:
This is not really safe -- the `Complete` call could be from a different goroutine from the one that is writing to `buf`, so you must synchronize. And just making `buf` an atomic pointer isn't sufficient because it's not just the pointer to `bytes.Buffer` but also the contents of the buffer.

I think it would be safe to instead have this `Complete` call merely store the trace in the `wireWrapper` atomic pointer, and then delay computing the wire details -- from the trace _and_ the buffer -- until the call to `getWireDetails`.

I _think_ that would be thread-safe, since the call to `getWireDetails` should always happen at the end of the RPC, after the whole thing is done and the entire response body has been read. So at that point, there should be no other concurrent activity the handler could be racing with.

Reply:
Yeah this actually also allows me to return an `error`, which in turn enables propagating this error back to the surface. I like this approach much better.